src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs

```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
+            {
+                List<Post> deserializedList = JsonSerializer.Deserialize<List<Post>>(json, options);
+
+                string jsonSerialized = JsonSerializer.Serialize(deserializedList, options);
+                Assert.Equal(json, jsonSerialized);
+            }
+
+            // Async case.
+            using (var memoryStream = new MemoryStream())
+            {
+                await JsonSerializer.SerializeAsync(memoryStream, posts[0], options);
+                string jsonSerialized = Encoding.UTF8.GetString(memoryStream.ToArray());
+                Assert.Equal(json, jsonSerialized);
+
+                memoryStream.Position = 0;
+                List<Post> deserializedList = await JsonSerializer.DeserializeAsync<List<Post>>(memoryStream, options);
+            }
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(16)]
+        public static async Task DeepNestedJsonFileCircularDependencyTest(int depthFactor)
+        {
+            int length = 10 * depthFactor;
+            List<Post>[] posts = new List<Post>[length];
+            posts[0] = PopulateLargeObject(1000);
+            for (int i = 1; i < length; i++)
+            {
+                posts[i] = PopulateLargeObject(1);
+                posts[i - 1][0].Value.PrimaryTopic.RelatedTopics = posts[i];
+            }
+            posts[length - 1][0].Value.PrimaryTopic.RelatedTopics = posts[0];
+
+            JsonSerializerOptions options = new JsonSerializerOptions()
+            {
+                MaxDepth = depthFactor * 64,
+                IgnoreNullValues = true
+            };
+
+            Assert.Throws<JsonException> (() => JsonSerializer.Serialize(posts[0], options));
+
+            using (var memoryStream = new MemoryStream())
+            {
+                await Assert.ThrowsAsync<JsonException>(async () => await JsonSerializer.SerializeAsync(memoryStream, posts[0], options));
+            }
+        }
+
+
+        [Theory]
+        [InlineData(128)]
+        [InlineData(1024)]
+        [InlineData(4096)]
+        [InlineData(8192)]
+        [InlineData(16384)]
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            int thresholdSize = (int)(bufferSize * 0.9 - 2);
+            FlushThresholdTestClass serialaizeObject = new FlushThresholdTestClass(PopulateObjectWhichProduceJsonWithGivenSize(bufferSize));
+            List<object> list = new List<object>();
+            StringBuilder builder = new StringBuilder();
+            for (int i = 0; i < thresholdSize; i++)
+            {
+                builder.Append("a");
+            }
+            list.Add(builder.ToString());
+            serialaizeObject.StringProperty = builder.ToString();
+            list.Add(serialaizeObject);
+            JsonSerializerOptions options = new JsonSerializerOptions();
+            options.DefaultBufferSize = bufferSize;
+
+            string json = JsonSerializer.Serialize(list);
```

Comment:
I'm having trouble understanding what this test is testing.

* Given a size input, produce a list of that many `(int)1` values.
* Produce a string with approximately 90% as many ASCII letters in it as the List\<int\> has ints.
* Make an object which holds both the list and the string.
* Make a second list which has the string then the object.
* Serialize the second list

What aspect is being tested?  options.DefaultBufferSize is set, but if we're trying to measure that it has impact I don't see that being done here.

* Did we need to set DefaultBufferSize?
* If the list didn't have both the string and the object is the test substantially different?
* If the object didn't have both the string and the list of ints, is the test substantially different?

Based on the name, I'd expect something measuring that DefaultBufferSize is performing "as expected", but unless not setting it would throw (which doesn't seem to be the case, since the call with no options succeeds) I don't see how this test is testing it...  what would fail if the serializer never read that option at all?

If I'm completely missing what's being tested that's OK, but then we should come up with a combination of name changes and comments that make it apparent what's being tested and how it could fail.

Reply:
It is testing if serializer flushing buffer as expected for edge conditions like new object or array starts exactly at the buffer threshold (where buffer flush occurs), testing it with different buffer size => with different threshold size. It is covering following sentence from the issue description #39500 
![image](https://user-images.githubusercontent.com/45579687/63128114-018d0080-bf69-11e9-9bc8-b107000459aa.png)

Reply:
Added comment explaining the object structure and its purpose in the test

``` 
// bufferSize * 0.9 is the threshold size from codebase, substruct 2 for [" characters, then create a 
// string containing (threshold - 2) amount of char 'a' which when written into output buffer produces buffer 
// which size equal to or very close to threshold size, then adding the string to the list, then adding a big 
// object to the list which changes depth of written json and should cause buffer flush
```
Hope its make it clear now and answering your first question.

>  Did we need to set DefaultBufferSize?

Yes to test with different threshold
> If the list didn't have both the string and the object is the test substantially different?

Yes as they have their own role, the string produces threshold length buffer when written, next object should cause buffer flush and its should be written without any loss  
> If the object didn't have both the string and the list of ints, is the test substantially different?

Not much, intended to test writing big enough object after flush